client/src/entrypoints/contrib/typed_table_block/typed_table_block.js

```diff
@@ -466,30 +466,44 @@ export class TypedTableBlock {
 
   getState() {
     const state = {
-      columns: this.columns.map((column) => ({
-        type: column.blockDef.name,
-        heading: column.headingInput.value,
-      })),
+      columns: this.getColumnState(),
       rows: this.rows.map((row) => ({
         values: row.blocks.map((block) => block.getState()),
       })),
     };
     return state;
   }
 
+  getDuplicatedState() {
+    return {
+      columns: this.getColumnState(),
+      rows: this.rows.map((row) => ({
+        values: row.blocks.map((block) =>
+          block.getDuplicatedState === undefined
+            ? block.getState()
+            : block.getDuplicatedState(),
+        ),
+      })),
+    };
+  }
+
   getValue() {
     const value = {
-      columns: this.columns.map((column) => ({
-        type: column.blockDef.name,
-        heading: column.headingInput.value,
-      })),
+      columns: this.getColumnState(),
       rows: this.rows.map((row) => ({
         values: row.blocks.map((block) => block.getValue()),
       })),
     };
     return value;
   }
 
+  getColumnState() {
```

Comment:
Nice improvement here!`getColumnsState` (with an s) might be ever-so-slightly clearer though, since there are usually multiple columns

Reply:
Good point!

Reply:
Also, do we need a `getDuplicatedColumnState()`, or do columns not have ids?

Reply:
If I'm not mistaken, we don't need to handle columns any differently. `TypedTableBlock.columns` is a list of objects that contain metadata for the column, including:
- the block definition of the "type" that can be inserted into rows in this column; and
- an integer ID, the index of the column, assigned when the column is created.

The IDs on the row items are generated in a similar fashion. These column and row IDs don't appear to have a unique constraint as they aren't used, e.g. as part of a contentpath for comments. The reason that we need special handling for rows is that the block type for a given column may be a complex type (e.g. a `StreamBlock`), which we would need to recurse into instances of, to make sure `StreamChild`, `ListChild`, etc. instances don't have duplicate IDs when duplicating a block.

However, this code is new to me so I'm happy to be corrected.

Reply:
Regarding the name, I've gone with `getColumnStates` - that feels more natural to me.

Reply:
Thanks for clarifying there @jams2. Happy with all of that :)